segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorageManager.java

```diff
@@ -0,0 +1,741 @@
+     * 2. Remove them from the source metadata.
+     * 3. Update the target version to ensure CAS.
+     *
+     * @param segmentName the target segment.
+     * @param sourceSegment The segment to be merged to target.
+     * @param offset offset at which the merge happens.
+     * @return A completable future which completes once the concat operation is complete.
+     */
+    public void concat(String segmentName, String sourceSegment, long offset) throws StreamSegmentException {
+        List<CuratorOp> curatorOps;
+        try {
+        curatorOps = this.getZKOperationsForConcat(segmentName, sourceSegment, offset);
+            List<CuratorTransactionResult> results = zkClient.transaction()
+                                                             .forOperations(curatorOps);
+            LogStorage logStorage = getOrRetrieveStorageLedger(segmentName);
+            logStorage.incrementUpdateVersion();
+            /* Fence all the ledgers and add one at the end. */
+            fenceLedgersAndCreateOneAtEnd(segmentName, logStorage);
+            /** Delete metadata and cache for the source once the operation is complete. */
+            this.zkClient.delete().deletingChildrenIfNeeded().forPath(getZkPath(sourceSegment));
+            ledgers.remove(sourceSegment);
+        } catch (Exception exc) {
+            /** In case of exception, drop the corrupt cache. */
+            ledgers.remove(segmentName);
+            throw translateZKException(segmentName, exc);
+        }
+    }
+
+    /**
+     * Deletes a segment along with all the data and metadata.
+     * @param segmentName name of the segment to be deleted.
+     * @return A CompletableFuture which completes once the delete is complete.
+     */
+    public void delete(String segmentName) throws StreamSegmentException {
+        LogStorage ledger = this.getOrRetrieveStorageLedger(segmentName);
+        if (ledger.getContainerEpoch() > this.containerEpoch.get()) {
+            throw new CompletionException(new StorageNotPrimaryException(segmentName));
+        }
+        try {
+            this.zkClient.delete()
+                         .deletingChildrenIfNeeded()
+                         .forPath(getZkPath(segmentName));
+        } catch (Exception e) {
+           log.warn("Exception while deleting a segment {}", segmentName, e);
+
+        }
+        ledgers.remove(segmentName);
+        deleteAllLedgers(ledger);
+    }
+
+    void deleteAllLedgers(LogStorage ledger) {
+        synchronized (ledger) {
+            ledger.getDataMap().entrySet().stream().forEach(entry -> deleteLedger(entry.getValue().getLedgerHandle()));
+        }
+    }
+
+    //endregion
+
+
+    public LedgerData createLedgerAt(String streamSegmentName, int offset) {
+
+        log.info("Creating ledger for {} at {}", streamSegmentName, offset);
+        try {
+            LedgerHandle ledgerHandle = bookkeeper.createLedger(config.getBkEnsembleSize(),
```

Comment:
What about adding some "custom metadata" on BookKeeper (this feature is available since 4.5) in order to mark this ledger as belonging to the given stream ?

This additional information will be useful in order to prevent accidental deletion or access to the wrong data (you can double check the presence of such metadata before delete operations or read operations).

I have cases in which the BookKeeper cluster will be shared with other applications, in this case having metadata in BookKeeper side will ease tooling to distinguish Pravega ledgers from ledgers pertaining to other applications.
I can file a issue, the same comment applies to for DurableLog